File: Canvas/PhotoEditingCanvasBrushStrokeView.m
//  Created by Geoff Pado on 8/27/19.
//  Copyright © 2019 Cocoatype, LLC. All rights reserved.

#import "PhotoEditingCanvasBrushStrokeView.h"

#import "PhotoEditingCanvasView.h"

@interface PhotoEditingCanvasBrushStrokeView ()

@property (nonatomic, strong) PhotoEditingCanvasView *canvasView;
@property (nonatomic) BOOL canvasViewIsResetting;
@property (nonatomic, nullable, strong) UIBezierPath *currentPath;

@end

@implementation PhotoEditingCanvasBrushStrokeView

@synthesize currentPath;

- (instancetype)init {
    if ((self = [super initWithFrame:CGRectZero])) {
        self.translatesAutoresizingMaskIntoConstraints = NO;

        self.canvasView = [[PhotoEditingCanvasView alloc] init];
        self.canvasView.delegate = self;
        [self addSubview:self.canvasView];

        [NSLayoutConstraint activateConstraints:@[
            [self.canvasView.widthAnchor constraintEqualToAnchor:self.widthAnchor multiplier:1.0],
            [self.canvasView.heightAnchor constraintEqualToAnchor:self.heightAnchor multiplier:1.0],
            [self.canvasView.centerXAnchor constraintEqualToAnchor:self.centerXAnchor],
            [self.canvasView.centerYAnchor constraintEqualToAnchor:self.centerYAnchor]
        ]];
    }

    return self;
}

- (void)updateToolWithCurrentZoomScale:(CGFloat)currentZoomScale {
    [self.canvasView updateToolWithCurrentZoomScale:currentZoomScale];
}

#pragma mark PKCanvasViewDelegate

- (void)canvasViewDrawingDidChange:(PKCanvasView *)canvasView {
    if (self.canvasViewIsResetting) { return; }

    self.currentPath = [self pathFromDrawing:self.canvasView.drawing];
    [self sendActionsForControlEvents:UIControlEventTouchUpInside];

    self.canvasViewIsResetting = YES;
    self.canvasView.drawing = [PKDrawing new];
    self.canvasViewIsResetting = NO;
}

#pragma mark Path Manipulation

- (UIBezierPath *)newPath {
    UIBezierPath *newPath = [UIBezierPath bezierPath];
    newPath.lineCapStyle = kCGLineCapRound;
    newPath.lineJoinStyle = kCGLineJoinRound;
    newPath.lineWidth = self.canvasView.currentLineWidth;
    return newPath;
}

- (UIBezierPath *)pathFromDrawing:(PKDrawing *)drawing {
    NSArray<PKStroke *> *strokes = [drawing strokes];
    PKStroke *currentStroke = [strokes lastObject];
    PKStrokePath *strokePath = [currentStroke path];
    UIBezierPath *bezierPath = [self newPath];
    [bezierPath moveToPoint:[strokePath interpolatedLocationAt:0]];
    for (NSUInteger i = 1; i < [strokePath count]; i++) {
        [bezierPath addLineToPoint:[strokePath interpolatedLocationAt:i]];
    }
    return bezierPath;
}

#pragma mark Touch Handling

//- (void)touchesBegan:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event {
//    UITouch *touch = [touches anyObject];
//    if (touch == nil) { return; }
//
//    self.currentPath = [self newPath];
//    [self.currentPath moveToPoint:[touch locationInView:self]];
//}
//
//- (void)touchesMoved:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event {
//    UITouch *touch = [touches anyObject];
//    if (touch == nil) { return; }
//    [self.currentPath addLineToPoint:[touch locationInView:self]];
//}
//
//- (void)touchesEnded:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event {
//    [self sendActionsForControlEvents:UIControlEventTouchUpInside];
//    self.currentPath = nil;
//}
//
//- (void)touchesCancelled:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event {
//    self.currentPath = nil;
//}

@end

File: Modules/Legacy/Editing/Sources/Editing View/Workspace/PhotoEditingCanvasBrushStrokeView.swift
//  Created by Geoff Pado on 10/31/24.
//  Copyright © 2024 Cocoatype, LLC. All rights reserved.

import PencilKit
import UIKit

class PhotoEditingCanvasBrushStrokeView: UIControl, PhotoEditingBrushStrokeView, PKCanvasViewDelegate {
    var currentPath: UIBezierPath? = nil

    init() {
        super.init(frame: .zero)
        translatesAutoresizingMaskIntoConstraints = false

        canvasView.delegate = self
        addSubview(canvasView)

        NSLayoutConstraint.activate([
            canvasView.widthAnchor.constraint(equalTo: widthAnchor, multiplier: 1),
            canvasView.heightAnchor.constraint(equalTo: heightAnchor, multiplier: 1),
            canvasView.centerXAnchor.constraint(equalTo: centerXAnchor),
            canvasView.centerYAnchor.constraint(equalTo: centerYAnchor),
        ])
    }

    var color: UIColor {
        get { canvasView.color }
        set { canvasView.color = newValue }
    }

    func updateTool(currentZoomScale: CGFloat) {
        canvasView.updateTool(currentZoomScale: currentZoomScale)
    }

    // MARK: Undo/Redo

    override var undoManager: UndoManager? { nil }

    // MARK: PKCanvasViewDelegate

    func canvasViewDrawingDidChange(_ canvasView: PKCanvasView) {
        guard canvasViewIsResetting == false else { return }

        currentPath = path(from: canvasView.drawing)

        sendActions(for: .touchUpInside)

        canvasViewIsResetting = true
        canvasView.drawing = PKDrawing()
        canvasViewIsResetting = false
    }

    // MARK: Path Manipulation

    private func newPath() -> UIBezierPath {
        let newPath = UIBezierPath()
        newPath.lineCapStyle = .round
        newPath.lineJoinStyle = .round
        newPath.lineWidth = canvasView.currentLineWidth
        return newPath
    }

    private func path(from drawing: PKDrawing) -> UIBezierPath {
        let bezierPath = newPath()
        if let strokePath = drawing.strokes.last?.path {
            bezierPath.move(to: strokePath.interpolatedLocation(at: 0))
            for i in (1..<strokePath.count) {
                bezierPath.addLine(to: strokePath.interpolatedLocation(at: CGFloat(i)))
            }
        }

        return bezierPath
    }

    // MARK: Boilerplate

    @available(*, unavailable)
    required init(coder: NSCoder) {
        let typeName = NSStringFromClass(type(of: self))
        fatalError("\(typeName) does not implement init(coder:)")
    }

    private let canvasView = PhotoEditingCanvasView()
    private var canvasViewIsResetting = false
}

